git-config/tests/integration_tests/git_config_integration_tests.rs

```diff
@@ -674,6 +674,87 @@ mod from_paths_tests {
         );
     }
 
+    #[test]
+    fn include_if_with_gitdir() {
+        let dir = tempdir().unwrap();
+
+        let a_path = dir.path().join("a");
+        let b_path = dir.path().join("b");
+        let c_path = dir.path().join("c");
+        let d_path = dir.path().join("d");
+        let inside_a_x_path = dir.path().join("a").join("x").join(".git");
+
+        fs::write(
+            a_path.as_path(),
+            format!(
+                r#"
+            [core]
+              b = 1
+            [includeIf "gitdir:~/.git"]
```

Comment:
What does `git` do if the `gitdir` is `~/repo` but the passed in `git_dir` for comparison is `~/repo` or `~/repo/` (trailing slash). Does it normalize these paths to some extend? I would expect it, but never checked.

I am asking because this is the kind of 'hidden smartness' that probably can be implemented before globbing.

Reply:
GIT_DIR has to be valid or there's error

```
➜  /Users/svetlin  GIT_DIR=~/IdeaProjects/gitoxide/.git git config --get core.filemode

false
➜  /Users/svetlin  GIT_DIR=~/IdeaProjects/gitoxide git config --get core.filemode

➜  /Users/svetlin  GIT_DIR=~/IdeaProjects/gitoxide/ git config --get core.filemode

➜  /Users/svetlin
```

Reply:
No errors in the opposite case:

with
```
[includeIf "gitdir:~/IdeaProjects/gitoxide"]
        path = ~/tmp/custom-git.cfg
```
or 
```
[includeIf "gitdir:/non-existing"]
        path = ~/tmp/custom-git.cfg
```
filemode is not changed
```
➜  /Users/svetlin  GIT_DIR=~/IdeaProjects/gitoxide/.git git config --get core.filemode

true
```

with
```
[includeIf "gitdir:~/IdeaProjects/gitoxide/"]
        path = ~/tmp/custom-git.cfg
```
filemode is changed
```
➜  /Users/svetlin  GIT_DIR=~/IdeaProjects/gitoxide/.git git config --get core.filemode

false
```

Reply:
Thanks for sharing your results! Changing the `GIT_DIR` environment variable wasn't my intention, I rather wanted to know what happens if the paths aren't normalized like this:

```git-config
# trailing slashes seem to work, right?
[includeIf "gitdir:~/IdeaProjects/gitoxide/"]
        path = ~/tmp/custom-git.cfg

# what about using the .git suffix?
[includeIf "gitdir:~/IdeaProjects/gitoxide/.git"]
        path = ~/tmp/custom-git.cfg

# what happens in bare repositories? Those without a work tree? Then there is only the full `GIT_DIR` as valid directory.
[includeIf "gitdir:~/IdeaProjects/gitoxide.git"]
        path = ~/tmp/custom-git.cfg

# And again, bare with trailing slash.
[includeIf "gitdir:~/IdeaProjects/gitoxide.git/"]
        path = ~/tmp/custom-git.cfg

# Since `gitoxide.git` is bare, the following shouldn't match
[includeIf "gitdir:~/IdeaProjects"]
        path = ~/tmp/custom-git.cfg
```

Since the work tree directory works as well as the git dir, I think the context must know if the repository is bare or not, or better, pass the working tree directory separately to assure we don't do anything forbidden. Just some thoughts born out of ignorance, and I hope it makes some sense, too 😅.

Reply:
Thanks so much! Also interesting that it adds `**` automatically on trailing slash, but it's probably a brute-force way to support `/.git` suffixes. It's most certainly intended so people can say: include this file if it is in this parent directory, leading to every repository in it being able to include the file.

By the way, this could be implemented as a prefix-search as well, so no glob needed in that particular case.

Reply:
Oh, and I think we should definitely have tests for that, too. The existing test can probably be extended for this.